src/compute/src/compute_state/peek_stash.rs

```diff
@@ -0,0 +1,245 @@
+
+        loop {
+            let row = rows_rx.recv().await;
+            match row {
+                Some(Ok(rows)) => {
+                    assert!(
+                        rows.len() > 0,
+                        "the contract is that we only get non-empty batches"
+                    );
+
+                    for (row, diff) in rows {
+                        num_rows +=
+                            u64::from(NonZeroU64::try_from(diff).expect("diff fits into u64"));
+                        let diff: i64 = diff.into();
+
+                        batch_builder
+                            .add(&SourceData(Ok(row)), &(), &Timestamp::default(), &diff)
+                            .await
+                            .expect("invalid usage");
+                    }
+                }
+                Some(Err(err)) => return Ok(PeekResponse::Error(err)),
+                None => {
+                    break;
+                }
+            }
+        }
+
+        let batch = batch_builder.finish(upper).await.expect("invalid usage");
+
+        let stashed_response = StashedPeekResponse {
+            num_rows: u64::cast_from(num_rows),
+            encoded_size_bytes: batch.encoded_size_bytes(),
+            relation_desc,
+            shard_id,
+            batches: vec![batch.into_transmittable_batch()],
+            inline_rows: RowCollection::new(vec![], &[]),
+        };
+        let result = PeekResponse::Stashed(Box::new(stashed_response));
+        Ok(result)
+    }
+
+    /// Pumps rows from the [PeekResultIterator] to the async task, via our
+    /// `rows_tx`. Will pump at most `batch_size` rows in one batch, and at most
+    /// the given `num_batches` batches.
+    pub fn pump_rows(&mut self, mut num_batches: usize, batch_size: usize) {
+        while self.peek_iterator.is_some() || self.stashed_rows.is_some() {
+            let rows = if let Some(rows) = self.stashed_rows.take() {
+                Some(rows)
+            } else {
+                match self.peek_iterator.as_mut() {
+                    Some(row_iter) => {
+                        let rows: Result<Vec<_>, _> = row_iter.take(batch_size).collect();
+                        let rows = rows.map(|rows| if rows.is_empty() { None } else { Some(rows) });
+                        rows.transpose()
+                    }
+                    None => None,
+                }
+            };
+
+            if let Some(rows) = rows {
+                match self
+                    .rows_tx
+                    .as_mut()
+                    .expect("missing rows_tx")
+                    .try_send(rows)
+                {
+                    Ok(_) => {
+                        // All good!
+                    }
+                    Err(TrySendError::Full(rows)) => {
+                        // Need to stash the row and try again later
+                        let prev = self.stashed_rows.replace(rows);
+                        assert!(prev.is_none());
```

Comment:
Don't we need to make sure we break out of the loop here? Otherwise we'll go through it again and hit the assert here, no?

Reply:
If we can get rid of the `stashed_rows` entirely, this part will also become much simpler :)